### DISCUSSION TITLE
Make ThreadPool replaceable with own + introduce IThreadPool instead of SynchronizationContext

### DISCUSSION THREAD

**Discussion by sidristij** [dotnet/runtime#65363] [Ideas]
Now we cannot use own pools as default and name "SynchronizationContext" means nothing in pooling context. I mean, if somebody who trying to learn how to work with threads and threads pooling see SynchronizationContext, no one will say: "ahh, this is abstraction layer for accessing some ThreadPool, ok". No. SyncContext is some place.... where what? I should synchronize some context with who? 

1) I suggesting to add type IThreadPool instead of use SynchronizationContext
2) I understand that it is impossible to make ThreadPool non-static. So, I suggest to add method ThreadPool.ReplaceImplementation(IThreadPool pool); And for every method, which allows to run something in parallel, make additional parameter "IThreadPool targetPool = default" to pass concrete model of parallelism.

**Comment by Clockwork-Muse** [comment]
A thread pool and synchronization context do different things.

That aside, generally you don't want or need to create your own thread pool - usually you'd create and use a [`TaskScheduler`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskscheduler?view=net-6.0) instead.  This is a fairly advanced usage, though, are you sure you need to do this?

**Reply by sidristij** [reply]
Disagree with you. TaskScheduler is for Tasks. It can cover SynchronizationContext to work on _some_ thread pool. Which can be regular ThreadPool, user-defined threadpool or single thread threadpool: for example, on UI SynchronizationContext. SynchronizationContext is abstraction layer of threadpool. [TaskScheduler.FromCurrentSynchronizationContext](https://source.dot.net/#System.Private.CoreLib/TaskScheduler.cs,a8790c1e6086ae5d)
